Pod/Classes/PINAnimatedImage.m

```diff
@@ -0,0 +1,494 @@
+{
+  *totalDuration = 0;
+  Float32 *durations = (Float32 *)malloc(sizeof(Float32) * frameCount);
+  [memoryMap getBytes:&durations range:NSMakeRange(18, sizeof(Float32) * frameCount)];
+
+  for (NSUInteger idx = 0; idx < frameCount; idx++) {
+    *totalDuration += durations[idx];
+  }
+
+  return durations;
+}
+
+- (Float32 *)durations
+{
+  return self.sharedAnimatedImage.durations;
+}
+
+- (CFTimeInterval)totalDuration
+{
+  return self.sharedAnimatedImage.totalDuration;
+}
+
+- (size_t)loopCount
+{
+  return self.sharedAnimatedImage.loopCount;
+}
+
+- (size_t)frameCount
+{
+  return self.sharedAnimatedImage.frameCount;
+}
+
+- (size_t)width
+{
+  return self.sharedAnimatedImage.width;
+}
+
+- (size_t)height
+{
+  return self.sharedAnimatedImage.height;
+}
+
+- (PINAnimatedImageStatus)status
+{
+  __block PINAnimatedImageStatus status;
+  [_statusLock lockWithBlock:^{
+    status = _status;
+  }];
+  return status;
+}
+
+- (CGImageRef)imageAtIndex:(NSUInteger)index
+{
+  return [self imageAtIndex:index
+         inSharedImageFiles:self.sharedAnimatedImage.maps
+                      width:(UInt32)self.sharedAnimatedImage.width
+                     height:(UInt32)self.sharedAnimatedImage.height
+                 bitmapInfo:self.sharedAnimatedImage.bitmapInfo];
+}
+
+- (PINImage *)coverImage
+{
+  return self.sharedAnimatedImage.coverImage;
+}
+
+- (BOOL)coverImageReady
+{
+  return self.status == PINAnimatedImageStatusInfoProcessed || self.status == PINAnimatedImageStatusFirstFileProcessed || self.status == PINAnimatedImageStatusProcessed;
+}
+
+- (BOOL)playbackReady
+{
+  return self.status == PINAnimatedImageStatusProcessed || self.status == PINAnimatedImageStatusFirstFileProcessed;
+}
+
+- (void)clearAnimatedImageCache
+{
+  [_dataLock lockWithBlock:^{
+    _currentData = nil;
+    _nextData = nil;
+  }];
+}
+
+- (NSUInteger)frameInterval
+{
+  return MAX(self.minimumFrameInterval * kPINAnimatedImageDisplayRefreshRate, 1);
+}
+
+//Credit to FLAnimatedImage (https://github.com/Flipboard/FLAnimatedImage) for display link interval calculations
+- (NSTimeInterval)minimumFrameInterval
+{
+  const NSTimeInterval kGreatestCommonDivisorPrecision = 2.0 / kPINAnimatedImageMinimumDuration;
+  
+  // Scales the frame delays by `kGreatestCommonDivisorPrecision`
+  // then converts it to an UInteger for in order to calculate the GCD.
+  NSUInteger scaledGCD = lrint(self.durations[0] * kGreatestCommonDivisorPrecision);
+  for (NSUInteger durationIdx = 0; durationIdx < self.frameCount; durationIdx++) {
+    Float32 duration = self.durations[durationIdx];
+    scaledGCD = gcd(lrint(duration * kGreatestCommonDivisorPrecision), scaledGCD);
+  }
+  
+  // Reverse to scale to get the value back into seconds.
+  return (scaledGCD / kGreatestCommonDivisorPrecision);
+}
+
+//Credit to FLAnimatedImage (https://github.com/Flipboard/FLAnimatedImage) for display link interval calculations
+static NSUInteger gcd(NSUInteger a, NSUInteger b)
```

Comment:
const references to show how you kick ass at c++? (stops a copy of the input variables)